zrml/prediction-markets/src/migrations.rs

```diff
@@ -15,259 +15,328 @@
+                        resolved_outcome = OutcomeReport::Scalar(to_fixed_point(value));
+                    }
+                    market.resolved_outcome = Some(resolved_outcome);
+                }
+
+                // Transform disputes using the same key because both maps have the same key
+                // type and hasher.
+                let old_disputes =
+                    get_storage_value::<DisputesOf<T>>(PREDICTION_MARKETS, DISPUTES, &key);
+                let new_disputes = match old_disputes {
+                    Some(disputes_unwrapped) => BoundedVec::try_from(
+                        disputes_unwrapped
+                            .into_iter()
+                            .map(|mut dispute| {
+                                if let OutcomeReport::Scalar(value) = dispute.outcome {
+                                    dispute.outcome = OutcomeReport::Scalar(to_fixed_point(value));
+                                };
+                                dispute
+                            })
+                            .collect::<Vec<_>>(),
+                    )
+                    .ok(),
+                    None => None,
+                };
+                new_scalar_markets.push((key, market, new_disputes));
+            }
+        }
+
+        for (key, market, new_disputes) in new_scalar_markets {
+            if let Some(disputes_unwrapped) = new_disputes {
+                put_storage_value::<DisputesOf<T>>(
+                    PREDICTION_MARKETS,
+                    DISPUTES,
+                    &key,
+                    disputes_unwrapped,
+                );
+                total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+            }
+
+            put_storage_value(MARKET_COMMONS, MARKETS, &key, market);
             total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
         }
-        log::info!("Completed updates of markets");
+
+        StorageVersion::new(PREDICTION_MARKETS_NEXT_STORAGE_VERSION).put::<Pallet<T>>();
         utility::put_storage_version_of_market_commons_pallet(MARKET_COMMONS_NEXT_STORAGE_VERSION);
-        total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        total_weight = total_weight.saturating_add(T::DbWeight::get().writes(2));
+        log::info!("TransformScalarMarketsToFixedPoint: Done!");
         total_weight
     }
 
     #[cfg(feature = "try-runtime")]
     fn pre_upgrade() -> Result<(), &'static str> {
+        // Check that no saturation occurs.
+        for (market_id, market) in T::MarketCommons::market_iter().drain() {
+            if let MarketType::Scalar(range) = market.market_type {
+                assert!(
+                    range.end().checked_mul(BASE).is_some(),
+                    "TransformScalarMarketsToFixedPoint: Arithmetic overflow when transforming \
+                     market {:?}",
+                    market_id,
+                );
```

Comment:
Nice overflow check here. 🥇 Edge case checker medal

I don't think this hits, but I wonder what we should do if this would happen in an imperfect world.

Reply:
To be honest, this test (almost) doesn't really do anything useful because someone could create a market that saturates between us running the test and deploying the update. If the PR wasn't already reviewed, I would remove that test.

If this test did discover any markets that saturate, we would just go through with the migration anyways. The on-chain democracy can decide whether or not they think it's a good idea to do this. The effect on the market is straightforward: The scalar range is restricted, which is good for the longs and bad for the shorts.

But there's one thing that shouldn't happen: Our scalar resolution assumes that `range.start() != range.end()`. I don't the time look it up, but I'm assuming that if `range.start() == range.end()`, then LONG and SHORT redeem for 1 ZTG, which is bad.

Reply:
> because someone could create a market that saturates between us running the test and deploying the update

But all other existing markets are checked. 

> we would just go through with the migration anyways. 

Doesn't `pre_upgrade` run before the `on_runtime_upgrade` and stops the migration if one assertion fails? 

> The effect on the market is straightforward: The scalar range is restricted, which is good for the longs and bad for the shorts.

The upper bound is limited. In case of an overflow, wouldn't be prices lower then and therefore it might be better for shorts instead of longs? 

> but I'm assuming that if range.start() == range.end(), then LONG and SHORT redeem for 1 ZTG, which is bad.

OK, so we should check for that?

Reply:
I've already added a check for `range.start() != range.end()`. Should have had that from the start, instead of checking for saturation.

The `pre_upgrade`/`post_upgrade` methods are only part of the `OnRuntimeUpgrade` trait `try-runtime` is enabled. They are only meant for testing purposes.

Reply:
> The upper bound is limited. In case of an overflow, wouldn't be prices lower then and therefore it might be better for shorts instead of longs?

Not really, the price resolution is based on where the bounds of the range are. So let's say you own a SHORT token for a market of range `[100, 200]`. Let's say this market would resolve to 175. That means your SHORT token would be worth .25 ZTG. Now suppose the the range is cut to `[100, 150]`. Now the market still resolves to 175, but that's the same as the market resolving to `150`, which means SHORT is worthless.

On the other hand, LONG would have been worth .75 ZTG before the change and 1 ZTG after.

Reply:
Ok, I read about how scalar markets work and understand the math behind it now. Long is just the percentage from the start of the range to the regarded (maybe resolved or current) price on the range. Short is just 1 minus the Long price.

> So let's say you own a SHORT token for a market of range [100, 200]. Let's say this market would resolve to 175. That means your SHORT token would be worth .25 ZTG.

`long_price = (range_price - range_start) / (range_end - range_start)`

`long_price = (175 - 100) / (200 - 100) = 75 / 100 = 0.75`

`short_price = 1 - long_price = 1 - 0.75 = 0.25`

> Now suppose the the range is cut to [100, 150]. Now the market still resolves to 175, but that's the same as the market resolving to 150, which means SHORT is worthless.

I see you are right 👍  Thanks for the explanation again.

Reply:
This is very informative discussion for me. :pray: 
if range.start() == range.end() then wouldn't there be some divide by zero math error in out code base somewhere, as chris's formula above?